src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,82 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            {
+                var oldDoc = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                var newDoc = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                GetMappedTextChanges(newDoc, oldDoc, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes to each file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.fileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.textChange), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            void GetMappedTextChanges(CodeAnalysis.Document newDocument, CodeAnalysis.Document oldDocument, ArrayBuilder<(string, TextChange)> builder)
+            {
+                var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                (var mappedSpanResults, var textChanges) = _threadingContext.JoinableTaskFactory.Run<(ImmutableArray<MappedSpanResult>, ImmutableArray<TextChange>)>(async () =>
+                {
+                    var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, CancellationToken.None).ConfigureAwait(false)).ToImmutableArray();
+                    var mappedSpanResults = await mappingService.MapSpansAsync(oldDocument, textChanges.Select(tc => tc.Span), CancellationToken.None).ConfigureAwait(false);
+
+                    return (mappedSpanResults, textChanges);
+                });
+
+                Contract.ThrowIfFalse(mappedSpanResults.Length == textChanges.Length);
+
+                for (var i = 0; i < mappedSpanResults.Length; i++)
+                {
+                    if (ShouldIncludeMappedChange(mappedSpanResults[i], newDocument))
+                    {
+                        builder.Add((mappedSpanResults[i].FilePath, new TextChange(mappedSpanResults[i].Span, textChanges[i].NewText)));
+                    }
+                }
+            }
+
+            static bool ShouldIncludeMappedChange(MappedSpanResult result, CodeAnalysis.Document newDocument)
+            {
+                if (result.IsDefault)
+                {
+                    // There is no appropriate mapping for the span, do not make a change.
+                    return false;
+                }
+
+                var documentIdsForFilePath = newDocument.Project.Solution.GetDocumentIdsWithFilePath(result.FilePath);
+                // Only attempt to apply text changes from mapped files that are not present in the workspace.
```

Comment:
how can we even apply text changes to things not in teh workspace.  I'm not even certain what that even means.  If it's not in teh workspace... where/waht is it?  Is it on disk?  Is it in memory?  How does an edit get applied to it?

Reply:
So we have the generated razor .cs document in the workspace.  But the top level .razor file is not part of our workspace.  The .razor file is on disk and part of the solution, just not part of our workspace.  We can apply an edit the same as any other file that's part of the project using the invisible editor to open a textbuffer for the filepath.